### PR TITLE
Tests for loader

### PR DESCRIPTION
**Pull request by tylerjw**
Related: https://github.com/cps-org/cps/issues/52

### CODE REVIEW

File: src/cps/loader.cpp [hunk 1]
```diff
@@ -87,7 +87,7 @@ namespace cps::loader {
             if (str == "symbolic") {
                 return Type::symbolic;
             }
-            CPS_UNREACHABLE(fmt::format("Unknown component.type: `{}`", str).c_str());
+            return Type::unknown;
```
Comment (dcbaker) [inline src/cps/loader.cpp:90]: In retrospect this was an inappropriate place to use `CPS_UNREACHABLE`, it makes more sense in the case of the `enum -> string`, where we actually can have high confidence that the input is something expected and that hitting in the URNEACHABLE case is a developer bug.

That said, I'm somewhat uncomfortable with adding an enum type that is essentially an error type, my preference would be to either use an `expected` or `optional` to note that something has gone wrong.

**Review by dcbaker — CHANGES_REQUESTED** [review]
Nice. I've definitely left a lot of error cases unhandled, and I'm glad to see them getting fixed!

File: src/cps/loader.cpp (hunk 1 above)
Comment (tylerjw) [reply src/cps/loader.cpp:90]: I went back and forth on this until I read these lines from the spec:

> If the type is not recognized by the parser, the component shall be ignored. (Parsers are permitted to support additional types as a conforming extension.)

This is not an error as much as it an unrecognized type by this parser. The set of valid types is open.

Comment (tylerjw) [reply src/cps/loader.cpp:90]: Side bar on enum/switch/return. I've debated myself about whether unreachable is even a good idea in the enum-to-string case. In C++, it is trivial to construct an enum from an integer. I've had instances where I used this pattern, which led to confusing bugs. In cases where the switch/return is over an enum, I either abort at the end or change the function to return optional and make the function consumer choose how much they trust the input.

Comment (dcbaker) [reply src/cps/loader.cpp:90]: Okay, with that escape hatch of "additional types are allowed" then that seems reasonable.

**Review by dcbaker — APPROVED** [review]

File: src/cps/loader.cpp [hunk 2]
```diff
@@ -7,19 +7,22 @@
 #include "cps/error.hpp"
 #include "cps/utils.hpp"
 
+#include <algorithm>
```
Comment (bretbrownjr) [inline src/cps/loader.cpp:10]: NBD, but it's nice to keep the includes for each dependency grouped together. It's also nice to sort them from high level to low level with the in-project headers at the top and standard libraries at the bottom.

File: src/cps/loader.cpp [hunk 3]
```diff
@@ -184,20 +190,15 @@ namespace cps::loader {
         get_components(const Json::Value & parent, std::string_view parent_name, const std::string & name) {
             Json::Value compmap;
             if (!parent.isMember(name)) {
-                return tl::unexpected(fmt::format("Required field Components of {} is missing!", parent_name));
+                return tl::unexpected(fmt::format("Required field `components` of `{}` is missing!", parent_name));
             }
 
             std::unordered_map<std::string, Component> components{};
 
             // TODO: error handling for not an object
```
Comment (bretbrownjr) [inline src/cps/loader.cpp:198]: @dcbaker What error handling did you expect? Something different than constructing an unexpected object?

File: src/cps/loader.cpp [hunk 4]
```diff
@@ -206,23 +207,49 @@ namespace cps::loader {
...
+                    .location = std::move(location),
+                    .link_location = std::move(link_location),
+                    .require = std::move(require),
+                };
+            }
+
+            if (components.empty()) {
+                return tl::unexpected(fmt::format("Components field of `{}` is empty, but must "
+                                                  "have at least one component",
+                                                  parent_name));
```
Comment (bretbrownjr) [inline src/cps/loader.cpp:252]:
```suggestion
                return tl::unexpected(fmt::format("`{}` must have at least one component",
                                                  parent_name));
```
...note that I haven't checked if clang-tidy is happy with this formatting.

File: src/cps/loader.cpp [hunk 5]
```diff
@@ -261,33 +279,44 @@ namespace cps::loader {
...
-        file.open(path);
-
+    tl::expected<Package, std::string> load(std::istream & input_buffer, std::string const & filename) {
         Json::Value root;
-        file >> root;
+        try {
+            input_buffer >> root;
+        } catch (std::exception const & ex) {
+            return tl::make_unexpected(
+                fmt::format("Exception caught while parsing json for `{}.cps`\n{}", filename, ex.what()));
```
Comment (bretbrownjr) [inline src/cps/loader.cpp:288]: We'll want to take another pass at user-friendly diagnostics in a future PR I expect, but let's not hold up this PR on that. I expect a parsing error should look a lot more like a useful compiler error at some point. Like "expected a closing quote on `foobar.cps:42,41`" for instance.

**Review by bretbrownjr — APPROVED** [review]
Approved with comments and minor suggestions.

I agree with Dylan that this is a great update. Eventually it would be nice to be to have CPS file validation as a first-class feature of `cps-config`, but let's talk about that outside of this PR. It would be great for packaging systems to validate CPS in their binary packages as part of the packaging process and failing to accept a binary package if there were issues.

* I like the backticks to disambiguate special field names in messages to the user
* NBD, but I'd be OK not implementing jar support at all. I don't think that's worthwhile for an MVP.
* I like the returning of `Type::unknown` instead of aborting. It's future-proof and that's nice. We can't expect all the tools generating CPS to be using older versions of CPS compared to `cps-config`.
* Blowing up the jumbo parameter list for the `Component` and `Package` constructors into a series of local variables that get passed in later is a nice readability improvement. If this gets too much bigger, we might want to consider using a builder pattern, though.

**Comment by tylerjw** [comment]
Thank you for the kind reviews. I've pushed a couple of small clean-ups.

**Comment by tylerjw** [comment]
I rebased this change, I believe it is ready. @dcbaker

**Pull request merged**